geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java

```diff
@@ -52,17 +62,32 @@ public Long putUserAuth(UserAuthAttributes userAuthAttr) {
     return newId;
   }
 
-  public Long putSubject(Subject subject, long existingUniqueId) {
-    final Long newId;
+
+  public long putSubject(@NotNull Subject subject, long existingUniqueId) {
+    final long newId;
     if (existingUniqueId == 0 || existingUniqueId == NOT_A_USER_ID) {
       newId = getNextID();
     } else {
       newId = existingUniqueId;
     }
 
-    Subject oldSubject = uniqueIdVsSubject.put(newId, subject);
-    removeSubject(oldSubject);
-    logger.debug("Subject of {} added.", newId);
+    // we are saving all the subjects that's related to this uniqueId
+    // we cannot immediately log out the old subject of this userId because
+    // it might already be bound to another thread and doing operations. If
+    // we log out that subject immediately, that thread "authorize" would get null principal.
+    synchronized (this) {
+      CopyOnWriteArrayList<Subject> subjects;
+      if (!uniqueIdVsSubject.containsKey(newId)) {
+        logger.debug("Subject of {} added.", newId);
+        subjects = new CopyOnWriteArrayList<>();
+        uniqueIdVsSubject.put(newId, subjects);
+      } else {
+        logger.debug("Subject of {} replaced.", newId);
+        subjects = uniqueIdVsSubject.get(newId);
+      }
+      // always add the latest subject to the top of the list;
+      subjects.add(0, subject);
```

Comment:
This will shift every element in the list, which could be costly. When adding a new subject, it makes a copy of the array. At the same time, the other thread could call `subjects.get(0)` and get the old subject before the new one is added.

Reply:
> This will shift every element in the list, which could be costly. When adding a new subject, it makes a copy of the array. At the same time, the other thread could call `subjects.get(0)` and get the old subject before the new one is added.

I think that `subjects.get(0)` is only invoked in a synchronized method which has the same lock scope as the `synchronized(this){}` block so that should prevent getting and putting to clash. 
Maybe the `synchronized(this){}` block could be limited to encapsulate just the `subjects.add(0, subject);` statement?

Reply:
for performance concerns, after we fix multiple threads in the client could send in authentication requests, `putSubject` should only be called  multiple times with the same uniqueId when credentials expires, which is not that frequent. I think we ignore the cost issue.

put/get are synchronized on the object itself, so there should not be `get(0)` happening while other thread is doing the put.

Reply:
Why does the latest subject have to be inserted as the first element in the list? Could we instead add that subject to the end of the list, and then in `getSubject(final Long userId)` retrieve the last element in the list? This would be a simple solution to the performance issue that Jianxia described.